File: Python/Circular_Singly_Linked_List.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class circularSinglyLinkedList:
    def __init__(self):
        self.last = None

    def insertFirst(self):
        data = int(input("Enter the value of the node to be inserted : "))
        newNode = Node(data)
        if (self.last == None):
            self.last = newNode
            newNode.next = self.last
        else:
            head = self.last.next
            self.last.next = newNode
            newNode.next = head

        print("The Node has been inserted in the beginning of the list\n")

    def deleteFirst(self):
        if (self.last == None):
            print("The list is empty\n")
        elif (self.last.next == self.last):
            print("Removed first : ", self.last.data, "\n")
            self.last = None
        else:
            head = self.last.next
            print("Removed first : ", head.data, "\n")
            head = head.next
            self.last.next = head

    def insertLast(self):
        data = int(input("Enter the value of the node to be inserted : "))
        newNode = Node(data)
        if (self.last == None):
            self.last = newNode
            newNode.next = self.last
        else:
            head = self.last.next
            self.last.next = newNode
            newNode.next = head
            self.last = newNode
        print("The Node has been inserted in the end of the list\n")

    def deleteLast(self):
        if (self.last == None):
            print("The list is empty\n")
        elif (self.last.next == self.last):
            print("Removed last : ", self.last.data, "\n")
            self.last = None
        else:
            print("Removed last : ", self.last.data, "\n")
            head = self.last.next
            curr = head
            while curr.next.next != head:
                curr = curr.next
            self.last = curr
            self.last.next = head

    def search(self, item):
        if (self.last == None):
            print("List is empty\n")
        else:
            head = self.last.next
            curr = head
            flag = 0
            while curr.next != head:
                if (curr.data == item):
                    flag = 1
                curr = curr.next

            if curr.data == item:
                flag = 1

            if (flag):
                print("item was found\n")
            else:
                print("not found\n")

    def display(self):
        if (self.last == None):
            print("List is empty\n")
        else:
            head = self.last.next
            curr = head
            print("the elements of the list are:")
            while curr.next != head:
                print(curr.data, end=" ")
                curr = curr.next
            print(curr.data, "\n")


CSLL = circularSinglyLinkedList()
while True:
    print("Enter a option you want to perform ")
    op = int(
        input("1.Insert First \n2.Delete First \n3.Insert Last \n4.Delete Last \n5.search \n6.display \n7.exit \n"))
    if (op == 1):
        CSLL.insertFirst()
    elif (op == 2):
        CSLL.deleteFirst()
    elif (op == 3):
        CSLL.insertLast()
    elif (op == 4):
        CSLL.deleteLast()
    elif (op == 5):
        item = int(input("enter the item to be searched : "))
        CSLL.search(item)
    elif (op == 6):
        CSLL.display()
    elif (op == 7):
        exit()
    else:
        print("enter correct option")

"""
Implementation of Circular_singly_Linked_List

Sample output:

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
1
Enter the value of the node to be inserted : 1
The Node has been inserted in the beginning of the list

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
1
Enter the value of the node to be inserted : 2
The Node has been inserted in the beginning of the list

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
1
Enter the value of the node to be inserted : 3
The Node has been inserted in the beginning of the list

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
6
the elements of the list are:
3 2 1

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
3
Enter the value of the node to be inserted : 5
The Node has been inserted in the end of the list

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
6
the elements of the list are:
3 2 1 5

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
2
Removed first :  3

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
6
the elements of the list are:
2 1 5

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
4
Removed last :  5

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
6
the elements of the list are:
2 1

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
5
enter the item to be searched : 1
item was found

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
5
enter the item to be searched : 6
not found

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
2
Removed first :  2

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
4
Removed last :  1

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
2
The list is empty

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
4
The list is empty

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
6
List is empty

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
5
enter the item to be searched : 2
List is empty

Enter a option you want to perform
1.Insert First
2.Delete First
3.Insert Last
4.Delete Last
5.search
6.display
7.exit
7

Time Complexity:
----------------------------------------------
Insert First   : O(1)
Delete First   : O(1)
Insert Last    : O(1)
Delete Last    : O(n)
Search         : O(n)

Space Complexity: 
-----------------------------------------------
Circular-singly-Linked-List : O(n)


"""

File: C-Plus-Plus/CircularSinglyLinkedList.cpp
#include<iostream>
#include<conio.h>
#include<stdlib.h>
using namespace std;
struct node
{
    int data;
    node *next;
};
node *head=NULL;

node* createNode(int x)
{
    node *p=(node*)malloc(sizeof(node));
    p->data=x;
    p->next=NULL;
    return p;
}

void addNodeAtEnd()
{
    int x;
    cout<<"Enter the element:";
    cin>>x;
    node *p=createNode(x);
    if(head==NULL)
    {
        head=p;
        p->next=head;
    }
    else
    {
        node *ptr=head;
        while(ptr->next!=head)
            ptr=ptr->next;
        ptr->next=p;
        p->next=head;
    }
}

void traverse()
{
    node *p=head;
    if(head==NULL)
        cout<<"List is Empty!";
    else
    {
        cout<<p->data<<" ";
        p=p->next;
        while(p!=head)
        {
            cout<<p->data<<" ";
            p=p->next;
        }
    }
}

void deleteEndNode()
{
    if(head==NULL)
        cout<<"List is already empty!";
    else if(head->next==head)
    {
        cout<<"Value removed:"<<head->data;
        head=NULL;
    }
    else
    {
        node *ptr=head;
        while(ptr->next->next!=head)
            ptr=ptr->next;
        cout<<"Value removed:"<<ptr->next->data;
        ptr->next=head;
    }
}

void addNodeAtBeg()
{
    int x;
    cout<<"Enter an element:";
    cin>>x;
    node *p=createNode(x);
    if(head==NULL)
    {
        head=p;
        p->next=head;
    }
    else
    {
        node *ptr=head;
        while(ptr->next!=head)
            ptr=ptr->next;
        ptr->next=p;
        p->next=head;
        head=p;
    }
}

void deleteFirstNode()
{
    if(head==NULL)
        cout<<"The list is already empty!";
    else if(head->next==head)
    {
        cout<<"Element deleted is:"<<head->data;
        head=NULL;
    }
    else
    {
        node *p=head;
        while(p->next!=head)
            p=p->next;
        p->next=head->next;
        cout<<"Element deleted is:"<<head->data;
        head=head->next;
    }
}

int main()
{
    int choice;
    while(1)
    {
        system("cls");
        cout<<"1. Display List\n";
        cout<<"2. Add an element at end\n";
        cout<<"3. Add element at beginning\n";
        cout<<"4. Delete last element\n";
        cout<<"5. Delete First Element\n";
        cout<<"6. Exit\n";
        cout<<"Enter your choice:";
        cin>>choice;
        switch(choice)
        {
        case 1:
            traverse();
            break;
        case 2:
            addNodeAtEnd();
            cout<<"Value Inserted!!";
            break;
        case 3:
            addNodeAtBeg();
            cout<<"Value Inserted!!";
            break;
        case 4:
            deleteEndNode();
            break;
        case 5:
            deleteFirstNode();
            break;
        case 6:
            exit(0);
        default:
            cout<<"\nChoose valid option!!";
        }
        getch();
    }
}

/*
Sample I/O:

1. Display List
2. Add an element at end
3. Add element at beginning
4. Delete last element
5. Delete First Element
6. Exit
Enter your choice:2
Enter the element:3
Value Inserted!!

1. Display List
2. Add an element at end
3. Add element at beginning
4. Delete last element
5. Delete First Element
6. Exit
Enter your choice:3
Enter an element:2
Value Inserted!!

1. Display List
2. Add an element at end
3. Add element at beginning
4. Delete last element
5. Delete First Element
6. Exit
Enter your choice:1
2 3

1. Display List
2. Add an element at end
3. Add element at beginning
4. Delete last element
5. Delete First Element
6. Exit
Enter your choice:5
Element deleted is:2

1. Display List
2. Add an element at end
3. Add element at beginning
4. Delete last element
5. Delete First Element
6. Exit
Enter your choice:4
Value removed:3

1. Display List
2. Add an element at end
3. Add element at beginning
4. Delete last element
5. Delete First Element
6. Exit
Enter your choice:1
List is Empty!

Time complexity:O(n)
Space complexiy:O(1)

*/
